fix(ci): optimize workflow skip logic to reduce runner waste

webjunkie opened the pull request:
Optimize CI skip behavior to prevent runner allocation waste. Previously, jobs checked skip conditions at step level, which still allocated runners and ran expensive setup steps. Moving to job-level conditions prevents any runner allocation when jobs should skip.

## Why This Changed

**Old pattern (wasteful):**
```yaml
job:
  runs-on: expensive-runner
  steps:
    - if: needs.changes.outputs.should_run == 'true'
      run: expensive_setup
```
Result: Runner allocated, then all steps skipped. Waste of runner time and $.

**New pattern (efficient):**
```yaml
job:
  runs-on: expensive-runner
  if: needs.changes.outputs.should_run == 'true'
  steps:
    - run: expensive_setup
```
Result: Job never starts, no runner allocated. Zero waste.

## Why This Works for Required Checks

Required status checks are summary jobs (e.g., "Django Tests Pass") that:
- Run with `if: always()` regardless of upstream status
- Check if upstream jobs succeeded OR were skipped
- Pass when all upstreams are either success or skipped

GitHub has treated job-level skipped jobs as passing for required checks since 2021. The old comments about "required checks can't skip" were outdated.

## Workflows Optimized

- **ci-e2e-playwright.yml**: Container + playwright jobs (prevented Qt libs install on skip)
- **ci-frontend.yml**: Format, toolbar, typescript, jest jobs
- **ci-storybook.yml**: Build + visual-regression jobs
- **ci-nodejs.yml**: Lint, build, test jobs
- **ci-backend.yml**: Django job (depot runner)
- **ci-python.yml**: Code quality job
- **container-images-ci.yml**: Docker build job (depot runner)

## Exception: Rust Workflow

Rust keeps step-level conditions because individual matrix jobs ("Test Rust services (feature-flags)") are required checks. Job-level skip prevents job names from registering, blocking PRs.

## Impact

Every PR that doesn't touch certain file types now skips jobs completely:
- Backend-only PRs: Skip frontend, storybook, Node.js jobs
- Frontend-only PRs: Skip backend, Python jobs
- Workflow-only PRs: Skip everything (with test commits removing self-triggers)

Saves expensive depot runner allocation, setup time (Node, pnpm, Python, Rust toolchain, Qt libs, Docker builds).

## Test Commits

Includes temporary commits removing workflow self-triggers to verify skipped jobs pass required checks. Will revert before merge.

webjunkie committed:
fix(ci): optimize E2E playwright workflow skip logic

Stop allocating runners and running setup steps when E2E tests should be skipped.

Problems fixed:
- `container` and `playwright-on-container` jobs allocated runners even when skipped
- Qt libs installed (huge dependency) even when tests skipped
- Dozens of setup steps ran just to satisfy outdated "required checks" pattern

Changes:
- Add job-level `if` conditions to prevent runner allocation when shouldRun is false
- Add `if` condition to Qt libs installation step
- Remove redundant step-level conditions (job-level handles it now)
- Delete obsolete comments about required checks (GitHub has supported skipped required checks since 2021)

This saves minutes per PR and significant runner costs when changes don't affect E2E tests.

webjunkie committed:
fix(ci): optimize skip logic across all CI workflows

Stop allocating runners and running setup steps when CI jobs should be skipped.

Problems fixed across 5 workflows:
- Jobs allocated runners even when changes didn't affect them
- Setup steps (Node, pnpm, Python, Rust, Qt libs, etc.) ran unnecessarily
- Wasted CI time and runner costs on every PR

Changes:
- Add job-level `if` conditions to prevent runner allocation when changes don't match
- Remove redundant step-level conditions (job-level prevents entire job from running)
- Delete obsolete comments about required checks (GitHub has supported skipped required checks since 2021)

Workflows fixed:
- ci-e2e-playwright.yml: Stop allocating runners for container/playwright jobs when skipped, prevent Qt libs installation
- ci-frontend.yml: Prevent all frontend jobs from allocating runners when no frontend changes
- ci-storybook.yml: Prevent storybook build/visual-regression from allocating runners when skipped
- ci-nodejs.yml: Prevent Node.js jobs from allocating runners when no Node.js changes
- ci-rust.yml: Prevent Rust jobs from allocating runners when no Rust changes

Summary jobs with `if: always()` ensure required status checks still pass when jobs are skipped.

webjunkie committed:
test: temporarily remove workflow self-triggers to test skip behavior

Remove workflow file paths from change filters to verify that:
1. Jobs are properly skipped when no relevant files change
2. Skipped jobs pass required status checks
3. No runners are allocated for skipped jobs

This commit will be reverted after testing.

webjunkie committed:
fix(ci): accept skipped jobs as passing in summary checks

Summary jobs were failing when upstream jobs were skipped because they only
checked for result == 'success', not accounting for result == 'skipped'.

When a job is skipped due to job-level if conditions, GitHub reports the
result as 'skipped', not 'success'. Summary jobs need to treat both as passing.

Fixed in:
- Frontend Tests Pass
- Node.js Tests Pass
- Visual regression tests pass

webjunkie committed:
fix(ci): optimize backend workflow skip logic

Stop allocating runners and running setup steps when backend tests should be skipped.

Problems fixed:
- Django test matrix jobs allocated runners even when no backend changes
- First step ran "echo Skipping backend tests" on allocated runner
- All setup steps had individual if conditions (wasteful)

Changes:
- Add job-level `if: needs.changes.outputs.backend == 'true'` to django job
- Remove conditional runner allocation (was using ubuntu-latest vs depot)
- Remove "Skip backend tests" echo step and all step-level conditions
- Fix "Django Tests Pass" summary job to accept skipped result as passing

Context:
PR #15174 (2 years ago) used step-level conditions due to GitHub matrix job
naming issues with required checks. However, PostHog now uses summary jobs
("Django Tests Pass") as the required check, which solves that problem.
Job-level skipping is now safe and significantly more efficient.

webjunkie committed:
test: remove backend workflow self-trigger for skip testing

Add backend workflow to the temporary test commit that removes workflow
self-triggers. This ensures backend CI won't run on the next push, allowing
us to verify that skipped jobs pass required checks.

webjunkie committed:
fix(ci): revert Rust optimization and optimize Python/E2E

Rust CI cannot use job-level skip because individual matrix jobs are required
checks (not a summary job). Reverting to step-level conditions for Rust.

Also optimized:
- ci-python.yml: Add job-level if, remove step-level conditions
- ci-e2e-playwright.yml (playwright job): Add job-level if, remove skip step

Context: PR #15174's concern about matrix jobs as required checks applies to
Rust CI but not to workflows with summary jobs (Backend, Frontend, Node.js,
Storybook all use summary jobs as required checks).

webjunkie committed:
test: remove all workflow self-triggers for skip testing

Remove workflow file paths from all change filters to verify that:
1. Jobs are properly skipped when no relevant files change
2. Skipped jobs pass required status checks
3. No runners are allocated for skipped jobs

Workflows updated:
- ci-python.yml
- ci-e2e-playwright.yml (in addition to previous test commit changes)

This commit will be reverted after testing.

webjunkie committed:
perf(ci): optimize container-images-ci.yml skip behavior

- Add job-level if condition to posthog_build to prevent runner allocation when docker files unchanged
- Remove redundant step-level if conditions
- Fix incorrect job name (was 'Rust checks', should be 'Docker checks')

webjunkie committed:
test: remove E2E workflow self-trigger for skip testing

webjunkie committed:
fix(ci): remove redundant step-level conditions

Job-level if already prevents job execution, so checking the same condition
at step level is redundant. Keep only the cache-hit check.

pauldambra reviewed (approved):
very quick scan...

the reason things were the way the were was because github required jobs are really annoyingly implemented

so we had to run steps in order to require them
we moved to a single required summary job since that made it easier to require a fixed thing

i guess the unexpected consequence is we can now do this

the only "risk" i can think of is we end up with a required job that isn't being run and blocks PRs, but that's really easy to fix so 🚢 🚢 🚢 🚢 🚢  and fix forward if necessary

joshsny reviewed (approved):
Nice! 👌

webjunkie committed:
Merge branch 'master' into fix/ci-e2e-playwright-skip-optimization

webjunkie committed:
Revert "test: temporarily remove workflow self-triggers to test skip behavior"

This reverts commit 23bd762da120817953ee5c86d877a160b13a6812.

webjunkie committed:
Revert "test: remove E2E workflow self-trigger for skip testing"

This reverts commit 8bfba29267cbb7dda2a5bb6383aa2c9a44f6970c.

webjunkie committed:
Revert "test: remove backend workflow self-trigger for skip testing"

This reverts commit 29747ab7c1b8d27fef742bfbff03b1a9965f71bc.

webjunkie committed:
Revert "test: remove all workflow self-triggers for skip testing"

This reverts commit 4e34acce8a93600f1d2fff966336e210e6f3028c.

webjunkie committed:
fix: remove blank lines introduced by workflow optimization

Remove blank lines that were left behind after removing step-level if conditions.
